### ISSUE TITLE
Missing tzinfo in audit columns

### ISSUE THREAD

**Issue by rafalkrupinski** [peterschutt/sqladmin-litestar-plugin#22]
```py
import sqladmin_litestar_plugin
from advanced_alchemy.base import UUIDAuditBase
from advanced_alchemy.extensions.litestar import SQLAlchemyPlugin
from litestar import Litestar
from litestar.contrib.sqlalchemy.plugins import SQLAlchemyAsyncConfig
from sqladmin import ModelView
from sqlalchemy import Column, String
from sqlalchemy.ext.asyncio import create_async_engine

engine = create_async_engine("sqlite+aiosqlite:///example.db")


class Entity(UUIDAuditBase):
    my_column = Column(String(10))


class EntityAdmin(ModelView, model=Entity):
    pass


app = Litestar(
    plugins=(
        SQLAlchemyPlugin(config=SQLAlchemyAsyncConfig(engine_instance=engine, create_all=True)),
        sqladmin_litestar_plugin.SQLAdminPlugin(views=[EntityAdmin], engine=engine, base_url='/admin'),
    ),
)
```

When creating a new entity via the Admin, the audit fields (created at, updated_at) get pre-populated in the form, with the current time. When trying to save the new entity sqlalchemy raises an error about missing tzinfo in the entity

```shell
sqlalchemy.exc.StatementError: (builtins.TypeError) tzinfo is required
[SQL: INSERT INTO entity (my_column, id, created_at, updated_at) VALUES (?, ?, ?, ?)]
[parameters: [{'my_column': 'test', 'updated_at': datetime.datetime(2024, 7, 17, 14, 30, 28), 'created_at': datetime.datetime(2024, 7, 17, 14, 30, 28)}]]
```

**Comment by peterschutt** [comment]
Thanks for the report @rafalkrupinski - and apologies for my delay in responding.

The advanced alchemy lib has custom DateTime types that enforce that timezone info is attached (see [here][1]), and these types are used on `UUIDAuditBase`.

Here's one way we can ensure the types have tzinfo attached after the admin form is submitted:

```py
class EntityAdmin(ModelView, model=Entity):
    async def on_model_change(
        self, data: dict, *_: Any
    ) -> None:
        data["created_at"] = data["created_at"].replace(tzinfo=datetime.timezone.utc)
        data["updated_at"] = data["updated_at"].replace(tzinfo=datetime.timezone.utc)
```

You could also deal with this  on the model:

```py
class EntityAdmin(ModelView, model=Entity):
    async def on_model_change(
        self, data: dict, *_: Any
    ) -> None:
        data["created_at"] = data["created_at"].replace(tzinfo=datetime.timezone.utc)
        data["updated_at"] = data["updated_at"].replace(tzinfo=datetime.timezone.utc)
```

[1]: https://github.com/litestar-org/advanced-alchemy/blob/main/advanced_alchemy/types/datetime.py

**Comment by rafalkrupinski** [comment]
I guess anyone could use such a custom column type, and it would break sqladmin. Do you know a way to plug some adapter directly into sqladmin?

**Comment by peterschutt** [comment]
> I guess anyone could use such a custom column type

True, however I don't think there is a way that we can build a general solution for that. We can have custom support for Advanced Alchemy's `DateTimeUTC` types, however.

Have a look at https://github.com/peterschutt/sqladmin-litestar-plugin/pull/30, your example above would become:

```py
from __future__ import annotations

from advanced_alchemy.base import UUIDAuditBase
from advanced_alchemy.extensions.litestar import SQLAlchemyPlugin
from litestar import Litestar
from litestar.contrib.sqlalchemy.plugins import SQLAlchemyAsyncConfig
from sqlalchemy import Column, String
from sqlalchemy.ext.asyncio import create_async_engine

import sqladmin_litestar_plugin
from sqladmin_litestar_plugin.ext.advanced_alchemy import AuditModelView

engine = create_async_engine("sqlite+aiosqlite:///")


class Entity(UUIDAuditBase):
    my_column = Column(String(10))


class EntityAdmin(AuditModelView, model=Entity): ...


app = Litestar(
    plugins=(
        SQLAlchemyPlugin(config=SQLAlchemyAsyncConfig(engine_instance=engine, create_all=True)),
        sqladmin_litestar_plugin.SQLAdminPlugin(
            views=[EntityAdmin], engine=engine, base_url="/admin"
        ),
    ),
)
```

(Basically, just swap out `sqladmin.ModelView` when using one of the AA audit base types for `sqladmin_litestar_plugin.ext.advanced_alchemy.AuditModelView`.)

**Comment by rafalkrupinski** [comment]
I think that's it aminalaee/sqladmin#796

**Comment by peterschutt** [comment]
Do you think the `AuditModelView` type that I implemented is a neat-enough solution to this?

**Comment by rafalkrupinski** [comment]
> Maybe those could be upstreamed into this lib as this problem won't be unique to using sqladmin with Litestar.

True, it starts with datepickr, with no updates for two years and almost 700 issues on GH, it's clearly an abandonware. SQLAdmin should migrate away from it already.

> Do you think the AuditModelView type that I implemented is a neat-enough solution to this?

Looks good. That's the kind of adapter I was asking about.
I guess the form converter could be used also without AuditBase, for user-defined fields, do you think it's worth mentioning it in the readme too?

**Comment by peterschutt** [comment]
> I guess the form converter could be used also without AuditBase, for user-defined fields, do you think it's worth mentioning it in the readme too?

Yeah, good point, it will definitely work if users use the DateTimeUTC type directly.

**Comment by peterschutt** [comment]
Closed in #30

**Comment by peterschutt** [comment]
This is somewhat addressed upstream now, as a validator has been added to the audit base variants that will coerce naive datetimes into UTC for those columns.

However, the DateTimeUTCConverter type will still be useful if anyone uses the `DateTimeUTC` column type from advanced-alchemy directly on their models.
